libcodechecker/analyze/tu_collector.py

```diff
@@ -0,0 +1,257 @@
+                      "compiler...")
+        try:
+            # Change the original compiler to the compiler from the toolchain.
+            command[0] = toolchain_compiler
+            dependencies |= set(__gather_dependencies(command, build_dir))
+        except Exception as ex:
+            logging.debug("Couldn't create dependencies:")
+            logging.debug(str(ex))
+            error += str(ex)
+
+    return dependencies, error
+
+
+def add_sources_to_zip(zip_file, files):
+    """
+    This function adds source files to the ZIP file if those are not present
+    yet. The files will be placed to the "sources-root" directory under the ZIP
+    file.
+
+    zip_file -- A ZIP file name as a string.
+    files -- A file path or an iterable of file paths.
+    """
+    if isinstance(files, str):
+        files = [files]
+
+    with zipfile.ZipFile(zip_file, 'a') as archive:
+        for f in files:
+            archive_path = os.path.join('sources-root', f.lstrip(os.sep))
+
+            try:
+                archive.getinfo(archive_path)
+            except KeyError:
+                archive.write(f, archive_path, zipfile.ZIP_DEFLATED)
+            else:
+                logging.debug("'%s' is already in the ZIP file, won't add it "
+                              "again!", f)
+
+
+def zip_tu_files(zip_file, compilation_database, write_mode='w'):
+    """
+    Collects all files to a zip file which are required for the compilation of
+    the translation units described by the given compilation database.
+    If there are some files which couldn't be collected then a "no-sources"
+    file will contain the error message of its reason.
+    The function returns the set of files.
+
+    zip_file -- A file name or a file object.
+    compilation_database -- Either a path of the compilation database JSON file
+                            or a list of the parsed JSON.
+    write_mode -- The file opening mode of the zip_file. In case of 'a' the new
+                  files are appended to the existing zip file, in case of 'w'
+                  the files are added to a clean zip file.
+    """
+    if isinstance(compilation_database, basestring):
+        with open(compilation_database) as f:
+            compilation_database = json.load(f)
+
+    no_sources = 'no-sources'
+    tu_files = set()
+    error_messages = ''
+
+    for buildaction in compilation_database:
+        files, err = get_dependent_headers(
+            buildaction['command'],
+            buildaction['directory'])
+
+        tu_files |= files
+
+        if err:
+            error_messages += buildaction['file'] + '\n' \
+                + '-' * len(buildaction['file']) + '\n' + err + '\n'
+
+    if write_mode == 'a' and os.path.isfile(zip_file):
+        with zipfile.ZipFile(zip_file) as archive:
+            try:
+                archive.getinfo(no_sources)
+            except KeyError:
+                pass
```

Comment:
Do not we want to write out some warning or debug log message here?

Reply:
It is just a way of querying whether the .zip file contains this file or not. Both cases are absolutely normal, so it wouldn't carry any useful information printing it out.